# 1.3 挂起是如何工作的？
&emsp;&emsp;挂起函数是Kotlin协程的标志，挂起能力是最基本的特征，其他概念都建立在它之上，这就是为什么本章的目标是深入理解它的工作原理。挂起一个协程意味着在中途停止它，这类似于停止一个视频游戏——你在一个检查点(checkpoint)保存，关闭游戏，你和你的电脑可以集中精力做其他的事情。然后，过一段时间，你想继续，所以你再次打开游戏，从保存的检查点恢复，这样你就可以从之前离开的地方开始游戏。这是协程的一个类比。它们也可以被停止，当它们停止时，它们会返回一个***Continuation***，这就像游戏中的保存点——我们可以使用它从我们停止的地方继续。

## 恢复（resume, 也译作“中断后继续”）
&emsp;&emsp;让我们来探究一下，首先，我们需要一个协程，创建一个挂起的主函数是最简单的方式，这是我们的出发点。
```kotlin
suspend fun main() {
    println("Before")
    
    println("After")
}

// Before
// After
```
这是一个简单的程序，它将打印“Before”和“After”。如果我们在中间挂起那么将会发生什么呢？为了达到挂起的目的，我们可以使用协程标准库提供的suspendCoroutine函数[^1]。
```kotlin
suspend fun main() {
    println("Before")

    suspendCoroutine<Unit> { }

    println("After")
}

// Before
```
如果你调用上面的代码，你看不到“After”，程序也不会停止运行(因为我们的主函数永远不会结束)。协程在"Before"之后被挂起。我们的游戏停止了，永远不会恢复。怎么办:question:之前提到可以让程序恢复的Continuation在哪里？再看一下对suspendCoroutine的调用，注意它以一个lambda表达式({&emsp;&emsp;})结束，作为参数传递进去的函数会在挂起之前被调用，这个函数得到一个continuation作为参数。
```kotlin
suspend fun main() {
    println("Before")

    suspendCoroutine<Unit> { continuation ->
        println("Before too")
    }

    println("After")
}

// Before
// Before too
```
一个函数就地调用另一个函数并不是什么新鲜事，这类似于let、apply或uselines。suspendCoroutine函数也是这样设计的，这使得在挂起之前使用continuation成为可能，在suspendCoroutine之后调用的话就太晚了。因此，作为参数传递给suspendCoroutine函数的lambda表达式会在挂起之前被调用，这个lambda用于存储在哪里继续，或者计划后续是否要恢复。

我们可以使用它立即恢复:
```kotlin
suspend fun main() {
    println("Before")

    suspendCoroutine<Unit> { continuation ->
        continuation.resume(Unit)
    }

    println("After")
}

// Before
// After
```
注意“After”在例子中被打印出来了，因为我们在suspendCoroutine中调用了恢复。
> 从Kotlin 1.3开始，Continuation的定义发生了变化。1.3之前Continuation类中只有resumeWith方法，1.3之后标准库中增加了两个扩展函数resume和resumeWithException，实际上内部也是直接调用resumewith，它们都不需要额外导包。
```kotlin
public inline fun <T> Continuation<T>.resume(value: T): Unit =
    resumeWith(Result.success(value))

public inline fun <T> Continuation<T>.resumeWithException(exception: Throwable): Unit =
    resumeWith(Result.failure(exception))
```
我们也可以用它来启动一个新线程，休眠一段时间后然后恢复:
```kotlin
suspend fun main() {
    println("Before")

    suspendCoroutine<Unit> { continuation ->
        thread {
            Thread.sleep(1000)
            continuation.resume(Unit)
        }
    }

    println("After")
}

// Before
// (1 second delay)
// After
```
这是个重要的发现。注意，启动线程可以被抽取到一个函数里面，恢复操作可以放在回调中，这样continuation就可被lambda捕获，如下面的代码所示：
```kotlin
fun invokeAfterSecond(operation: () -> Unit) {
    thread {
        Thread.sleep(1000)
        operation.invoke()
    }
}

suspend fun main() {
    println("Before")

    suspendCoroutine<Unit> { continuation ->
        invokeAfterSecond {
            continuation.resume(Unit)
        }
    }

    println("After")
}

// Before
// (1 second delay)
// After
```
这种方式是可行的，但是它创建了线程然后停止活动一秒钟后线程就被结束了，这是不必要的，线程可并不便宜(开销大)，为什么要浪费它们呢？一个更好的方式是设置一个闹钟，在jvm上，我们可以使用ScheduledExecutorService，我们可以设置它在一定的时间调用continuation.resume(Unit)。
```kotlin
private val executor = Executors.newSingleThreadScheduledExecutor {
    Thread(it, "scheduler").apply { isDaemon = true }
}

suspend fun main() {
    println("Before")

    suspendCoroutine<Unit> { continuation ->
        executor.schedule({
            continuation.resume(Unit)
        }, 1000, TimeUnit.MILLISECONDS)
    }

    println("After")
}

// Before
// (1 second delay)
// After
```
挂起一定的时间这似乎是个很有用的功能，让我们把它提取到一个函数中，命名为delay。
```kotlin
private val executor = Executors.newSingleThreadScheduledExecutor {
    Thread(it, "scheduler").apply { isDaemon = true }
}

suspend fun delay(time: Long) {
    suspendCoroutine<Unit> { continuation ->
        executor.schedule({
            continuation.resume(Unit)
        }, time, TimeUnit.MILLISECONDS)
    }
}

suspend fun main() {
    println("Before")

    delay(1000)

    println("After")
}

// Before
// (1 second delay)
// After
```
executor仍然使用的是线程，但是它只有一个线程来为所有协程执行delay函数，这比每次我们需要等待一段时间的时候而去阻塞一个线程好多了。这正是Kotlin协程库之前实现delay的方式，现在的实现更加复杂，主要为了支持测试，不过基本思路还是一样的。

## 带着值恢复
你可能会关心的一件事是，为什么我们要将Unit传递给resume函数。你可能还想知道为什么我使用Unit作为suspendCoroutine的类型参数。事实上这两个参数为同一类型并不是巧合，lambda参数中Continuation的泛型类型、suspendCoroutine函数的最终返回结果类型都是Unit。
```kotlin
val ret: Unit = suspendCoroutine<Unit> { cont: Continuation<Unit> ->
    cont.resume(Unit)
}
```
当我们调用suspendCoroutine，我们可以指定continuation返回的类型，当我们调用resume时也要传入相应类型的对象。
```kotlin
suspend fun main() {
    val i: Int = suspendCoroutine<Int> { continuation ->
        continuation.resume(42)
    }
    println(i) // 42

    val str: String = suspendCoroutine<String> { continuation ->
        continuation.resume("Some text")
    }
    println(str) // Some text

    val b: Boolean = suspendCoroutine<Boolean> { continuation ->
        continuation.resume(true)
    }
    println(b) // true
}
```
这和之前的游戏类比有点不同，我不知道有哪一款游戏，可以在保存点恢复时放入一些东西到游戏的某个地方(除非你作弊去搜索下一个挑战，否则这都算新知识)。不过这对协程来说是讲得通的，在等待数据的时候我们经常被挂起，例如，我们从api中获取网络响应，这是个很常见的场景。你的线程执行业务逻辑直到它到达某个点需要去获取数据，接着它会让你的网络库交付这些数据，如果没用协程，这个线程就要停下来等待，这是个巨大的浪费——线程是非常昂贵的，特别是一个非常重要的线程，例如Android中的主线程。如果用协程，它(指协程)只是挂起，给网络库一个恢复指令“一旦你获取到了数据，把它们放在resume函数就行”，然后线程就可以去做其他事情，一旦数据达到了，这个线程就可以在协程挂起的点恢复。

为了，让我们看看怎么挂起直到接收到一些数据，在例子中，我们用了一个外部已经实现好了的回调函数requestUser。
```kotlin
suspend fun main() {
    println("Before")
    val user = suspendCoroutine<User> { continuation ->
        requestUser { user ->
            continuation.resume(user)
        }
    }
    println(user)
    println("After")
}

// Before
// (1 second delay)
// User(name=Test)
// After
```
直接调用suspendCoroutine并不方便，我们更希望有一个挂起的函数，所以我们可以自己提取。
```kotlin
suspend fun requestUser(): User {
    return suspendCoroutine<User> { continuation ->
        requestUser { user ->
            continuation.resume(user)
        }
    }
}

suspend fun main() {
    println("Before")
    val user = requestUser()
    println(user)
    println("After")
}
```
目前，许多流行的库已经支持挂起功能，如Retrofit或Room，这就是为什么我们很少需要在挂起函数中使用回调函数。如果你有这样的需求，我建议使用suspendCancellableCoroutine(而不是suspendCoroutine)，这将在*Cancellation*章节解释。
```kotlin
suspend fun requestUser(): User {
    return suspendCancellableCoroutine<User> { continuation ->
        requestUser { user ->
            continuation.resume(user)
        }
    }
}
```
## 带着异常恢复
我们调用的每个函数都可能返回某些值或抛出异常，这同样适用于suspendCoroutine。当resume被调用时，它会返回作为参数传入的数据。当resumeWithException被调用时，作为参数传递的异常从概念上讲是从挂起点抛出的。
```kotlin
suspend fun main() {
    try {
        suspendCoroutine<Unit> { continuation ->
            continuation.resumeWithException(MyException())
        }
    } catch (e: MyException) {
        println("Caught!")
    }
}

// Caught!
```
这种机制可用于各种各样的问题，例如，显示网络异常。
```kotlin
suspend fun requestUser(): User {
    return suspendCancellableCoroutine<User> { continuation ->
        requestUser { resp ->
            if (resp.isSuccessful) {
                continuation.resume(resp.data)
            } else {
                val e = ApiException(
                    resp.code,
                    resp.message
                )
                continuation.resumeWithException(e)
            }
        }
    }
}

suspend fun requestNews(): News {
    return suspendCancellableCoroutine<News> { continuation ->
        requestNews(
            onSuccess = { news -> continuation.resume(news) },
            OnErrorAction = { e -> continuation.resumeWithException(e) }
        )
    }
}
```
## 挂起一个协程，而不是一个函数
这里需要强调的一件事是，我们挂起的是协程，而不是函数。假设我们将一个函数存储在某个变量中，并尝试在函数调用后恢复它。
```kotlin
// ❌千万别这么做！！！
var continuation: Continuation<Unit>? = null

suspend fun suspendAndSetContinuation() {
    suspendCoroutine<Unit> { cont ->
        continuation = cont
    }
}

suspend fun main() {
    println("Before")

    suspendAndSetContinuation()
    continuation?.resume(Unit)

    println("After")
}
// Before
```
这毫无意义，这相当于停止游戏，并计划在游戏的后面部分恢复游戏，恢复永远不会被调用(译注：只能在停止的地方恢复，停止之后的地方在没恢复之前根本没法执行到)。你只会看到“Before”，而你的程序永远不会结束，恢复需要发生在另一个线程或另一个协程中，为了展示这一点，我们可以设置另一个协程在一秒钟后恢复。
```kotlin
// ❗不要这么做，可能存在内存泄漏
var continuation: Continuation<Unit>? = null

suspend fun suspendAndSetContinuation() {
    suspendCoroutine<Unit> { cont ->
        continuation = cont
    }
}

suspend fun main() = coroutineScope {
    println("Before")

    launch { // 新开了一个子协程，1秒后触发父协程恢复
        delay(1000)
        continuation?.resume(Unit)
    }

    suspendAndSetContinuation()
    
    // 译注：如果上方的launch放在这里，结果和上面的例子一样，因为在父协程挂起之后，后面的代码必须恢复才能执行的到

    println("After")
}

// Before
// (1 second delay)
// After
```
## 总结
&emsp;&emsp;我希望你现在已经从用户的角度对挂起是如何工作的有了一个清晰的认知，这一点很重要，我们将会看到它(指挂起)贯穿本书，它也是实用的，因为现在你可以把回调函数变成挂起函数。如果你像我一样想知道挂起具体是如何工作的，你可能还然想知道它是如何实现的，如果你对此感到好奇，这些都会在下一章介绍。如果你觉没必要知道，就跳过下一章，毕竟不是特别实用，只是展示了Kotlin协程的魅力。

[^1]: suspendCoroutine函数内部直接调用了suspendCoroutineUninterceptedOrReturn，这是一个原生函数，具体实现并不可见，因为它是和平台相关的([Intrinsic](https://juejin.cn/post/6863119536200253454))。
